File: Samples/Buttons/Unity/Assets/NoesisGUI/Samples/Buttons/ElementExtensions.cs
#if UNITY_5_3_OR_NEWER
#define NOESIS
using Noesis;
#else
using System;
using System.Windows;
using System.Windows.Controls;
using System.Windows.Controls.Primitives;
#endif

namespace Buttons
{
    public class ElementExtensions: DependencyObject
    {
        public ElementExtensions()
        {
        }

        #region FocusOnLoaded
        public static DependencyProperty FocusOnLoadedProperty = DependencyProperty.RegisterAttached(
            "FocusOnLoaded", typeof(bool), typeof(ElementExtensions), new PropertyMetadata(false, OnFocusOnLoadedChanged));

        public static bool GetFocusOnLoaded(DependencyObject d)
        {
            return (bool)d.GetValue(FocusOnLoadedProperty);
        }

        public static void SetFocusOnLoaded(DependencyObject d, bool value)
        {
            d.SetValue(FocusOnLoadedProperty, value);
        }

        private static void OnFocusOnLoadedChanged(DependencyObject d, DependencyPropertyChangedEventArgs e)
        {
            FrameworkElement element = d as FrameworkElement;
            if (element != null && (bool)e.NewValue == true)
            {
                element.Loaded += (s, a) => element.Focus();
            }
        }
        #endregion

        #region FocusOnHover
        public static DependencyProperty FocusOnHoverProperty = DependencyProperty.RegisterAttached(
            "FocusOnHover", typeof(bool), typeof(ElementExtensions), new PropertyMetadata(false, OnFocusOnHoverChanged));

        public static bool GetFocusOnHover(DependencyObject d)
        {
            return (bool)d.GetValue(FocusOnHoverProperty);
        }

        public static void SetFocusOnHover(DependencyObject d, bool value)
        {
            d.SetValue(FocusOnHoverProperty, value);
        }

        private static void OnFocusOnHoverChanged(DependencyObject d, DependencyPropertyChangedEventArgs e)
        {
            UIElement element = d as UIElement;
            if (element != null && (bool)e.NewValue == true)
            {
                element.Focus();
            }
        }
        #endregion

        #region SelectOnHover
        public static DependencyProperty SelectOnHoverProperty = DependencyProperty.RegisterAttached(
            "SelectOnHover", typeof(bool), typeof(ElementExtensions), new PropertyMetadata(false, OnSelectOnHoverChanged));

        public static bool GetSelectOnHover(DependencyObject d)
        {
            return (bool)d.GetValue(SelectOnHoverProperty);
        }

        public static void SetSelectOnHover(DependencyObject d, bool value)
        {
            d.SetValue(SelectOnHoverProperty, value);
        }

        private static void OnSelectOnHoverChanged(DependencyObject d, DependencyPropertyChangedEventArgs e)
        {
            UIElement element = d as UIElement;
            if (element != null && (bool)e.NewValue == true)
            {
                Selector.SetIsSelected(element, true);
            }
        }
        #endregion

        #region SelectAllOnFocus
        public static DependencyProperty SelectAllOnFocusProperty = DependencyProperty.RegisterAttached(
            "SelectAllOnFocus", typeof(bool), typeof(ElementExtensions), new PropertyMetadata(false, OnSelectAllOnFocusChanged));

        public static bool GetSelectAllOnFocus(DependencyObject d)
        {
            return (bool)d.GetValue(SelectOnHoverProperty);
        }

        public static void SetSelectAllOnFocus(DependencyObject d, bool value)
        {
            d.SetValue(SelectOnHoverProperty, value);
        }

        private static void OnSelectAllOnFocusChanged(DependencyObject d, DependencyPropertyChangedEventArgs e)
        {
            UIElement element = d as UIElement;
            if (element != null && (bool)e.NewValue == true)
            {
                element.GotFocus += (sender, args) =>
                {
                    if (sender is TextBox)
                    {
                        ((TextBox)sender).SelectAll();
                    }
                    else if (sender is PasswordBox)
                    {
                        ((PasswordBox)sender).SelectAll();
                    }
                };
            }
        }
        #endregion
    }
}

File: Samples/Buttons/C++/Src/ElementExtensions.cpp
////////////////////////////////////////////////////////////////////////////////////////////////////
// NoesisGUI - http://www.noesisengine.com
// Copyright (c) 2013 Noesis Technologies S.L. All Rights Reserved.
////////////////////////////////////////////////////////////////////////////////////////////////////


#include "ElementExtensions.h"

#include <NsCore/ReflectionImplement.h>
#include <NsCore/TypeId.h>
#include <NsGui/FrameworkElement.h>
#include <NsGui/UIElementData.h>
#include <NsGui/PropertyMetadata.h>
#include <NsGui/Selector.h>
#include <NsGui/PasswordBox.h>
#include <NsGui/TextBox.h>


using namespace Buttons;
using namespace Noesis;


////////////////////////////////////////////////////////////////////////////////////////////////////
bool ElementExtensions::GetFocusOnLoaded(DependencyObject* d)
{
    return d->GetValue<bool>(FocusOnLoadedProperty);
}

////////////////////////////////////////////////////////////////////////////////////////////////////
void ElementExtensions::SetFocusOnLoaded(DependencyObject* d, bool value)
{
    d->SetValue<bool>(FocusOnLoadedProperty, value);
}

////////////////////////////////////////////////////////////////////////////////////////////////////
bool ElementExtensions::GetFocusOnHover(DependencyObject* d)
{
    return d->GetValue<bool>(FocusOnHoverProperty);
}

////////////////////////////////////////////////////////////////////////////////////////////////////
void ElementExtensions::SetFocusOnHover(DependencyObject* d, bool value)
{
    d->SetValue<bool>(FocusOnHoverProperty, value);
}

////////////////////////////////////////////////////////////////////////////////////////////////////
bool ElementExtensions::GetSelectOnHover(DependencyObject* d)
{
    return d->GetValue<bool>(SelectOnHoverProperty);
}

////////////////////////////////////////////////////////////////////////////////////////////////////
void ElementExtensions::SetSelectOnHover(DependencyObject* d, bool value)
{
    d->SetValue<bool>(SelectOnHoverProperty, value);
}

////////////////////////////////////////////////////////////////////////////////////////////////////
bool ElementExtensions::GetSelectAllOnFocus(DependencyObject* d)
{
    return d->GetValue<bool>(SelectAllOnFocusProperty);
}

////////////////////////////////////////////////////////////////////////////////////////////////////
void ElementExtensions::SetSelectAllOnFocus(DependencyObject* d, bool value)
{
    d->SetValue<bool>(SelectAllOnFocusProperty, value);
}

////////////////////////////////////////////////////////////////////////////////////////////////////
void ElementExtensions::OnFocusOnLoadedChanged(DependencyObject* d,
        const DependencyPropertyChangedEventArgs& e)
{
    FrameworkElement* element = NsDynamicCast<FrameworkElement*>(d);
    if (element != 0 && *(bool*)e.newValue)
    {
        element->Loaded() += [](BaseComponent* sender, const RoutedEventArgs&)
        {
            FrameworkElement* element = (FrameworkElement*)sender;
            element->Focus();
        };
    }
}

////////////////////////////////////////////////////////////////////////////////////////////////////
void ElementExtensions::OnFocusOnHoverChanged(DependencyObject* d,
        const DependencyPropertyChangedEventArgs& e)
{
    FrameworkElement* element = NsDynamicCast<FrameworkElement*>(d);
    if (element != 0 && *(bool*)e.newValue)
    {
        element->Focus();
    }
}

////////////////////////////////////////////////////////////////////////////////////////////////////
void ElementExtensions::OnSelectOnHoverChanged(DependencyObject* d,
        const DependencyPropertyChangedEventArgs& e)
{
    FrameworkElement* element = NsDynamicCast<FrameworkElement*>(d);
    if (element != 0 && *(bool*)e.newValue)
    {
        Selector::SetIsSelected(element, true);
    }
}

////////////////////////////////////////////////////////////////////////////////////////////////////
void ElementExtensions::OnSelectAllOnFocusChanged(DependencyObject* d,
    const DependencyPropertyChangedEventArgs& e)
{
    UIElement* element = NsDynamicCast<UIElement*>(d);
    if (element != 0 && *(bool*)e.newValue)
    {
        element->GotFocus() += [](BaseComponent* sender, const RoutedEventArgs&)
        {
            PasswordBox* passwordBox = NsDynamicCast<PasswordBox*>(sender);
            if (passwordBox != 0)
            {
                passwordBox->SelectAll();
                return;
            }

            TextBox* textBox = NsDynamicCast<TextBox*>(sender);
            if (textBox != 0)
            {
                textBox->SelectAll();
                return;
            }
        };
    }
}

////////////////////////////////////////////////////////////////////////////////////////////////////
NS_IMPLEMENT_REFLECTION(Buttons::ElementExtensions)
{
    NsMeta<TypeId>("Buttons.ElementExtensions");

    Ptr<UIElementData> data = NsMeta<UIElementData>(TypeOf<SelfClass>());
    data->RegisterProperty<bool>(FocusOnLoadedProperty, "FocusOnLoaded",
        PropertyMetadata::Create(false, &ElementExtensions::OnFocusOnLoadedChanged));
    data->RegisterProperty<bool>(FocusOnHoverProperty, "FocusOnHover",
        PropertyMetadata::Create(false, &ElementExtensions::OnFocusOnHoverChanged));
    data->RegisterProperty<bool>(SelectOnHoverProperty, "SelectOnHover",
        PropertyMetadata::Create(false, &ElementExtensions::OnSelectOnHoverChanged));
    data->RegisterProperty<bool>(SelectAllOnFocusProperty, "SelectAllOnFocus",
        PropertyMetadata::Create(false, &ElementExtensions::OnSelectAllOnFocusChanged));
}

////////////////////////////////////////////////////////////////////////////////////////////////////
const DependencyProperty* ElementExtensions::FocusOnLoadedProperty;
const DependencyProperty* ElementExtensions::FocusOnHoverProperty;
const DependencyProperty* ElementExtensions::SelectOnHoverProperty;
const DependencyProperty* ElementExtensions::SelectAllOnFocusProperty;
